Refactor fetchIOUReportByID

luacmartins opened the pull request:
### Details
Removes `fetchIOUReportByID` since that data is returned in `OpenReport` and `ReadOldestAction`

cc @Julesssss @mountiny 

### Fixed Issues
$ https://github.com/Expensify/Expensify/issues/249195  

### Tests
Should be tested with the [Web-E PR](https://github.com/Expensify/Web-Expensify/pull/35818)

1. From account A, send an IOU request to account B
2. In account B, pay the IOU request
3. Log into account A and verify that the paid IOU preview loads 
4. From account A, send 50+ messages to account B
5. Send another IOU request
6. From account B, pay the latest IOU request
7. Log out and back into account A
8. Verify that you see the 2nd paid IOU preview without any loading spinners
9. Scroll back in the chat until you see the 1st IOU preview.
10. Verify that you see the 1st paid IOU preview without any loading spinners

- [x] Verify that no errors appear in the JS console

### Offline tests
N/A

### QA Steps
Same as test steps

- [ ] Verify that no errors appear in the JS console

### PR Author Checklist
<!--
This is a checklist for PR authors. Please make sure to complete all tasks and check them off once you do, or else your PR will not be merged!
-->

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] iOS / native
    - [x] Android / native
    - [x] iOS / Safari
    - [x] Android / Chrome
    - [x] MacOS / Chrome
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is correct English and approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

https://user-images.githubusercontent.com/22219519/207720454-d9c72c3f-fc3a-44a0-ae65-3ed6babb640f.mov

</details>

<details>
<summary>Mobile Web - Chrome</summary>

https://user-images.githubusercontent.com/22219519/207720527-7a72535f-8293-47f1-9ca8-c82da2b3e115.mov

</details>

<details>
<summary>Mobile Web - Safari</summary>

https://user-images.githubusercontent.com/22219519/207720549-2bb15aff-a8db-4109-ab72-0ef2b2901150.mov

</details>

<details>
<summary>Desktop</summary>

https://user-images.githubusercontent.com/22219519/207720567-b2da235e-9184-4dc6-93e4-3e31ef1a7f00.mov

</details>

<details>
<summary>iOS</summary>

https://user-images.githubusercontent.com/22219519/207720591-ff10ca08-b636-443e-8501-8c3a595a0f64.mov

</details>

<details>
<summary>Android</summary>

https://user-images.githubusercontent.com/22219519/207720613-c377016c-25d3-4129-9a8d-fa8381adf21a.mov

</details>

luacmartins committed:
rename fetchIOUReportByID to openIOUPreview

luacmartins committed:
rm setLocalIOUReportData

luacmartins committed:
rm unused imports

luacmartins committed:
rm fetchIOUReport

luacmartins committed:
convert ioupreview to class component

luacmartins committed:
rm export of getSimplifiedIOUReport

luacmartins commented on src/components/ReportActionItem/IOUPreview.js:
```diff
@@ -101,121 +101,124 @@ const defaultProps = {
     isHovered: false,
 };
 
-const IOUPreview = (props) => {
```
I converted `IOUPreview` to a class component to call `Report.openIOUPreview` only once in `componentDidMount`. This prevents that command from being called 6 times as we were doing before.

luacmartins committed:
rm openIOUPreview

luacmartins committed:
revert class component

luacmartins committed:
rm iouReportID prop

luacmartins committed:
rm isLoadingReport

luacmartins committed:
rm unused import

luacmartins commented on src/libs/actions/Report.js:
```diff
@@ -181,76 +178,6 @@ function getSimplifiedIOUReport(reportData, chatReportID) {
-            // This is expected and we return early as no further processing can be done.
-            return;
-        }
-        return getSimplifiedIOUReport(iouReportData, chatReportID);
```
This is still used in [syncChatAndIOUReports](https://github.com/Expensify/App/blob/dca62bd52329de8a26572eb60ca6fbb9ab793f63/src/libs/actions/Report.js#L942). We can remove it once `PayWithWallet` is refactored

mountiny replied:
🎉

luacmartins committed:
revert prop removal and suppress lint error

luacmartins commented:
@mollfpr please hold on testing this until the API changes are in staging!

mountiny reviewed (commented):
Nice! code looks good to me, gonna test tomorrow hopefully

mollfpr commented:
## Reviewer Checklist

- [x] I have verified the author checklist is complete (all boxes are checked off).
- [ ] I verified the correct issue is linked in the `### Fixed Issues` section above
- [ ] I verified testing steps are clear and they cover the changes made in this PR
    - [ ] I verified the steps for local testing are in the `Tests` section
    - [ ] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [ ] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [ ] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [ ] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I verified tests pass on **all platforms** & I tested again on:
    - [ ] iOS / native
    - [ ] Android / native
    - [ ] iOS / Safari
    - [ ] Android / Chrome
    - [ ] MacOS / Chrome
    - [ ] MacOS / Desktop
- [ ] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [ ] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [ ] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [ ] I verified that comments were added to code that is not self explanatory
    - [ ] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [ ] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [ ] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [ ] I verified any copy / text that was added to the app is correct English and approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [ ] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [ ] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [ ] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [ ] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [ ] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [ ] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [ ] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [ ] If a new component is created I verified that:
    - [ ] A similar component doesn't exist in the codebase
    - [ ] All props are defined accurately and each prop has a `/** comment above it */`
    - [ ] The file is named correctly
    - [ ] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [ ] The only data being stored in the state is data necessary for rendering and nothing else
    - [ ] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [ ] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [ ] All JSX used for rendering exists in the render method
    - [ ] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [ ] If any new file was added I verified that:
    - [ ] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [ ] If a new CSS style is added I verified that:
    - [ ] A similar style doesn't already exist
    - [ ] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [ ] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [ ] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [ ] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [ ] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>Mobile Web - Chrome</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>Mobile Web - Safari</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>Desktop</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>iOS</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>Android</summary>

<!-- add screenshots or videos here -->

</details>

mountiny reviewed (dismissed):
Tested well for me locally

Julesssss commented on src/components/ReportActionItem/IOUPreview.js:
```diff
@@ -33,6 +32,7 @@ const propTypes = {
     onPayButtonPressed: PropTypes.func,
 
     /** The active IOUReport, used for Onyx subscription */
+    // eslint-disable-next-line react/no-unused-prop-types
     iouReportID: PropTypes.string.isRequired,
```
I can't see this being used now. Why can't we remove it, am I missing something?

mollfpr replied:
I also have the same thought, but `iouReportID` is used to get the report from Onyx.
https://github.com/Expensify/App/blob/a414d1a30c1726ae6dcfd44d4f28e1ccaf3b7cdf/src/components/ReportActionItem/IOUPreview.js#L227-L229

Julesssss replied:
So the IOUPrevew presumably is missing the iouReport data if we remove it, but I can't find the exact reason why. Could you please point to it so I can resolve this thread.

mountiny replied:
I think what @mollfpr linked there is where it is used. the HOC used the id to fetch the iouReport and that is then used in the component.

mollfpr replied:
`iouReport` data is used to show the IOU data no? Like email, total, currency, and hasOutstandingIOU.

mountiny replied:
yes

Julesssss replied:
Yeah, I was looking for prop usage and didn't spot this. Thanks

Julesssss reviewed (dismissed):
Looks good, I just left a question because I can't figure out why THIS component needs the iouReport prop.

mollfpr commented:
Can we clean this up? `iou` props are no longer used in `IOUPreview` and we get the IOU report data in the component. 
https://github.com/Expensify/App/blob/a414d1a30c1726ae6dcfd44d4f28e1ccaf3b7cdf/src/pages/iou/IOUDetailsModal.js#L149-L150

luacmartins committed:
rm unused iou prop

luacmartins commented:
@mollfpr removed the iou prop!

mollfpr commented:
Just a curiosity, in detail, mentioned that now we using `OpenIOUPreview` but I don't see any `OpenIOUPreview` is called.

luacmartins commented:
@mollfpr nice catch! That was the first solution that I implemented and I forgot to update the description to reflect the new improved solution. I updated it now, so it's correct!

mollfpr reviewed (dismissed):
LGTM 👍

Will test it when the API hit the staging!

mountiny reviewed (approved):
Nice catch with the unused prop! 

@mollfpr This is ready to be tested against Staging!

mollfpr commented:
Updated [checklist](https://github.com/Expensify/App/pull/13556#issuecomment-1352466408) and test well on Web, Desktop, iOS, and Android.

I can't go through the mWeb testing because my IP is still not whitelisted. I believe the result will be the same...

mountiny commented:
@mollfpr Thank you very much, I think that sounds fine to me.

@luacmartins There are some merge conflicts 😢

mountiny reviewed (dismissed):
I have tested this in mWeb locally and it was a pass, although even locally sending messages quickly is super slow for some reason.

https://user-images.githubusercontent.com/36083550/208447216-42a7c337-669f-40e6-979d-1305f965d595.mp4

luacmartins committed:
resolve conflicts

luacmartins commented:
Since Web-E is not in prod yet and `PayWithWallet` might be merged today, I'll hold this PR so we can also remove [this](https://github.com/Expensify/App/pull/13556#discussion_r1048949368)

mountiny reviewed (approved):
@luacmartins Thank you, this is no ton hold anymore as the Web PR hit production.

I think we can make that follow up clean up in a new PR just to also keep the concerns separate, maybe we can wait for the PayWithWallet to get to production and confirm there are no regressions.

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/3740870731) to staging by @mountiny in version: 1.2.42-1 🚀

 platform | result 
 ---|--- 
🤖 android 🤖|success ✅ 
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅ 
🕸 web 🕸|success ✅

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/3753320419) to production by @yuwenmemon in version: 1.2.42-2 🚀

 platform | result 
 ---|--- 
🤖 android 🤖|success ✅ 
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅ 
🕸 web 🕸|success ✅

mollfpr commented:
@luacmartins Could you pull the BZ of the issue to help wrap this up? Thank you!

mountiny commented:
@mollfpr handling it in the issue, sorry realized it is internal, but we are on track cc @puneetlath

mallenexpensify commented:
Handling payment via this E/app issue https://github.com/Expensify/App/issues/13896
